core/src/main/scala/org/apache/spark/deploy/history/HistoryPage.scala

```diff
@@ -34,18 +37,31 @@ private[spark] class HistoryPage(parent: HistoryServer) extends WebUIPage("") {
     val requestedIncomplete =
       Option(request.getParameter("showIncomplete")).getOrElse("false").toBoolean
 
-    val allApps = parent.getApplicationList().filter(_.completed != requestedIncomplete)
-    val actualFirst = if (requestedFirst < allApps.size) requestedFirst else 0
-    val apps = allApps.slice(actualFirst, Math.min(actualFirst + pageSize, allApps.size))
-
+    val allCompletedAppsNAttempts = 
+        parent.getApplicationList().filter(_.completed != requestedIncomplete)
+    val (hasAttemptInfo, appToAttemptMap)  = getApplicationLevelList(allCompletedAppsNAttempts)
```

Comment:
Just to make my suggestions a little more concrete, I'm suggesting something like this:

- `ApplicationHistoryInfo` becomes `ApplicationAttemptInfo`
- Add a new `ApplicationHistoryInfo` as follows:

    private[spark] class ApplicationHistoryInfo(id: String, attempts: Seq[ApplicationAttemptInfo])

- `getListing()` remains unchanged, but note the data it returns is different now.

So now you have a list of applications being returned, and each application has information about all its attempts. Now it's much easier to slice that list and choose what to render in `HistoryPage`. It's a list, so it integrates nicely with `UIUtils.listingTable`, unlike the map I suggested before.